### PR TITLE
8277830: jextract generated code should not depend on Java layouts

### PR DESCRIPTION
**Pull request by mcimadamore**
This patch tweaks jextract generation, so that a new constant class is emitted, namely `Constant$root`, which contains the definition of all basic C primitive types. These are declared with the correct alignment constraints, depending on platforms.

Every other layout created by jextract will depend on these shared layouts constants, so that jextract code doesn't depend on alignment choices in `ValueLayout` constants.

In the future, we should probably generalize this approach, so that e.g. layouts are emitted only once, and then referred to afterwards - not just for primitive types, but also for everything else (struct, typedef, etc.). This would minimize the amount of layouts we store in the constant classes.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issue
 * [JDK-8277830](https://bugs.openjdk.java.net/browse/JDK-8277830): jextract generated code should not depend on Java layouts

### Reviewers
 * [Athijegannathan Sundararajan](https://openjdk.java.net/census#sundar) (@sundararajana - Committer)
 * [Jorn Vernee](https://openjdk.java.net/census#jvernee) (@JornVernee - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/panama-foreign pull/620/head:pull/620` \
`$ git checkout pull/620`

Update a local copy of the PR: \
`$ git checkout pull/620` \
`$ git pull https://git.openjdk.java.net/panama-foreign pull/620/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 620`

View PR using the GUI difftool: \
`$ git pr show -t 620`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/panama-foreign/pull/620.diff">https://git.openjdk.java.net/panama-foreign/pull/620.diff</a>

</details>

### CODE REVIEW

**Comment by mcimadamore** [comment]
I've tested this on some samples I have on Linux - would probably be wise to also test on Windows/Mac.

**Review by sundararajana — APPROVED** [review]
approved.

build/test fine on Mac. samples run fine on Mac as well

File: src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/ToplevelBuilder.java [hunk 1]
```diff
@@ -207,6 +216,74 @@ String build() {
...
+        }
+
+        @Override
+        protected String primitiveLayoutString(ValueLayout vl) {
+            if (vl.carrier() == boolean.class) {
+                return "JAVA_BOOLEAN";
+            } else if (vl.carrier() == char.class) {
+                return "JAVA_CHAR.withBitAlignment(" + vl.bitAlignment() + ")";
+            } else if (vl.carrier() == byte.class) {
+                return "JAVA_BYTE";
```
[the remark this answers is not in the document]
Comment (JornVernee) [reply src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/ToplevelBuilder.java:245]: Ah I guess this code is only called from the methods below.

[the remark this answers is not in the document]
Comment (mcimadamore) [reply src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/ToplevelBuilder.java:245]: yeah - this is just for the primitive layout that get added "by default".

[the remark this answers is not in the document]
Comment (mcimadamore) [reply src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/ToplevelBuilder.java:245]: Note also that we do not really have a way to deal with hyper (or under) aligned types (e.g. pragma pack). See:
https://bugs.openjdk.java.net/browse/JDK-8259832

**Review by JornVernee — APPROVED** [review]

**Pull request closed without merging**